README
The USBWall project
-------------------

The goal of the usbwall project is to add a complete USB mass-storage filtering system in the Linux
kernel.
Through filtering we would like to support:

1) USB Storage white-listing, permitting to allow only specific pre-registered usb devices. Other
devices are simply unaccessible from any userspace tools such as mount.
2) USB Storage fine grain filtering, specifying which action can be done for each allowed device,
such as read, write, execute, setu(g)id root execution, etc. This function needs to be integrated
in a complete security suite which is not supported by now.
3) Simple listing management through graphical user interface.
4) High level API definition through a userspace library, allowing to update or modify the module
interface without impacting userspace software. Only the abstraction library has to be updated.

The USBWall module
------------------
This is the source repository of the USBWall module.
This sources contain only the Linux kernel module. The library and the currently being implemented
applications can be found on github:
https://github.com/LACSC/

The module currently support whitelisting through usb mass storage devices injection. By now, there
is no automatic key injection at startup. This should be done by the user using the startup scripts
of his distribution

Limitations
-----------
In order to be functionnal, the usb_storage module has to be compiled as a module (not statically)
this is the case in all major distributions for desktop), and the usbwall module has to be loaded
_before_ usb_storage. In the other case, the usb_storage module will take the hand on the device
first and the filtering will not be supported.

This limitation will be corrected in the future.


ChangeLog
commit af90e8149862e79762cf8af223aea505ef9e7d00
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 17:12:04 2012 +0100

    corrected invalid kfree on ioctl error cases.

commit ebb1054246b56e292669d1e100952cecb9284a2f
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 15:49:17 2012 +0100

    support for usbwall dev done.

commit 6e1079770f2f0868e4a5b9533a706c3f0f44cbcf
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 15:15:41 2012 +0100

    add proper AUTHORS, INSTALL and README file

commit 505f899ca032d77281b53b28e17bb17d702eeee8
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 11:21:04 2012 +0100

    updated ChangeLog

commit 0967739b1d27c93ccbe7d7c206b1dd61d522b266
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 11:20:25 2012 +0100

    add depmod (native install). Add support for chrdev ioctl (add and del key). Add single concurrent access to device (simple method).

commit 5f885f7f64797938cee3e07e8587fcafb0bca3f5
Merge: 57a2ad2 8398563
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 10:36:35 2012 +0100

    merging trace

commit 57a2ad24c40daf298080ed3cca305a01cfba2445
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Feb 2 10:29:18 2012 +0100

    add control car device /dev/usball support. Need to add some real content in open/ioctl/close and single access support

commit 8398563f5bd9a9ae74fd33efb9927cd97852437e
Author: David Fernandes <dfernand@ece.fr>
Date:   Fri Jan 27 11:58:58 2012 +0100

    update ChangeLog

commit 41e05a511efb49a54f63f6bd3ba9eca41d4dbcde
Author: David Fernandes <dfernand@ece.fr>
Date:   Fri Jan 27 11:56:43 2012 +0100

    update USBwall informations

commit 0a20543f5053e678737aa9e9b384ae2233d19535
Author: David Fernandes <dfernand@ece.fr>
Date:   Fri Jan 27 11:08:21 2012 +0100

    update ChangeLog

commit 7b8b56f67e9d0a8a5e500eff1374352246f949f1
Author: David Fernandes <dfernand@ece.fr>
Date:   Fri Jan 27 11:07:30 2012 +0100

    name update

commit efe745b5fbbde52a6c9ef256717d023138ef37bb
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Jan 26 14:02:28 2012 +0100

    add DBG_LEVEL support (through dbglevel module option. This permit to manage correctly the module log level

commit 31bb1039155d55177b71318c560e3a8d27fa2af8
Author: Philippe Thierry <thierry@ece.fr>
Date:   Thu Jan 26 10:52:45 2012 +0100

    purge misplaced char in trace submodule.

commit bb6c7cba8448164aef054c9c35f184bfe4cc597c
Author: David Fernandes <dfernand@ece.fr>
Date:   Wed Jan 25 15:52:53 2012 +0100

    update ChangeLog

commit f60a4728f1a74d7184f4f95848ea55d802f25427
Author: David Fernandes <dfernand@ece.fr>
Date:   Wed Jan 25 15:52:08 2012 +0100

    status, read key list

commit 7b922f735617334f12bb7fc1c6056edfcb385052
Author: David Fernandes <dfernand@ece.fr>
Date:   Tue Jan 24 19:00:40 2012 +0100

    update ChangeLog

commit c3ceee9ec4a076ac0381de06601224ede931db9f
Author: David Fernandes <dfernand@ece.fr>
Date:   Tue Jan 24 18:52:54 2012 +0100

    fixed bug rmmod, add tests in functions "authorize" and "del keyinfo", adding license file

commit 6b93085ac5961045340a582052aaab0352b4c521
Author: David Fernandes <dfernand@ece.fr>
Date:   Tue Jan 24 10:58:39 2012 +0100

    add & del key bug fixed

commit 4e99dcaeb1834c8a740729befa8d4859108fb4fe
Merge: 662b23a ba8fb71
Author: David Fernandes <dfernand@ece.fr>
Date:   Mon Jan 23 17:40:50 2012 +0100

    Merge branch 'master' of ../tmp/usbwall

commit 662b23add6b7b5c1fbfa1ef85921728073549324
Author: David Fernandes <dfernand@ece.fr>
Date:   Mon Jan 23 17:37:06 2012 +0100

    update to merge

commit 55dd7f9c3d12bf981dc77757438e68412f9791f9
Author: David Fernandes <dfernand@ece.fr>
Date:   Mon Jan 23 17:15:47 2012 +0100

    update ChangLog

commit e9ef3f9629f9d826a63bc0bc535a5023cc190efe
Author: David Fernandes <dfernand@ece.fr>
Date:   Mon Jan 23 17:09:34 2012 +0100

    bug in keylist.c file when we use the "list_for_each_entry" function

commit beba93a6a85107ecc778ba357c981c0f96422819
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 17:17:56 2012 +0100

    update changlog

commit 542bd82109dd725aaca6c1201c29a63b66c10abf
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 17:16:48 2012 +0100

    add keylist_info.h

commit ba8fb7133b30cca975f71dc2724304fd1138138d
Merge: d99ebee 27adea0
Author: Philippe Thierry <phil@reseau-libre.net>
Date:   Sun Jan 22 17:10:39 2012 +0100

    Merge branch 'master' of git://github.com/DavidFernandes/USBWall

commit 27adea0cab5175378f9a7e90bfa9ea260af922d7
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 16:26:52 2012 +0100

    update ChangeLog

commit beeb6d051b3e227921ada1f2094c807e5a6fc869
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 16:24:14 2012 +0100

    rename "mass_storage_info" to "usbwall_token_info"
    update fonction key_del
    add intanal structure for list

commit d99ebee38c9146b7ee78df571bc59be0fc37a3f6
Merge: 92c0fb2 d04a7ef
Author: Philippe Thierry <phil@reseau-libre.net>
Date:   Sun Jan 22 15:35:14 2012 +0100

    Merge branch 'master' of git://github.com/DavidFernandes/USBWall

commit d04a7ef00e16a85d398708226dd08e40557573a0
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 14:51:30 2012 +0100

    remove list_head

commit 92c0fb252a9c3499454a1401aa91b51ad571e9e1
Merge: 40aa211 5421abd
Author: Philippe Thierry <phil@reseau-libre.net>
Date:   Sun Jan 22 12:24:24 2012 +0100

    merge from last David update. Add a node for numeric coded release for valiation check in libusbwall init.

commit 5421abd7751c69db8bd5cba41ee5ef86513243b6
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 11:25:00 2012 +0100

    update ChangeLog

commit bb1d5a9872971cefaa7aa1b4df06e00aeabf32d3
Author: David Fernandes <dfernand@ece.fr>
Date:   Sun Jan 22 11:22:04 2012 +0100

    deletion deprrecated files

commit 8ca12142bf5357f6e44326fc3a75d379ff76d5ff
Author: David Fernandes <dfernand@ece.fr>
Date:   Sat Jan 21 15:12:33 2012 +0100


src/procfs_iface.c
/*
** File procfs_iface.c for project usbwall
**
** LACSC - ECE PARIS Engineering school 
**
** Made by Philippe THIERRY
** Login   <phil@reseau-libre.net>
**
** Started on  jeu. 19 janv. 2012 10:34:39 CET Philippe THIERRY
**
** Copyright (C) 2009 - Philippe THIERRY
**
** This program is free software; you can redistribute it and/or
** modify it under the terms of the GNU Lesser General Public License
** as published by the Free Software Foundation; either version 2
** of the License, or (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/
/*
** \file procfs_iface.c
**
** Brief documentation on one line
** Detailed documentation
** On multiple lines...
**
** \author Philippe THIERRY
**
*/

#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/init.h>
#include <linux/proc_fs.h>
#include <linux/sched.h>
#include <linux/version.h>
#include <linux/slab.h>
#include <asm/uaccess.h>
#include "procfs_iface.h"
#include "trace.h"
#include "usbwall.h"
#include "keylist.h"
#include "keylist_info.h"

#define USBWALL_PROC_STATUS_BUFFER_SIZE 256
#define USBWALL_PROC_RELEASE_BUFFER_SIZE 16

static struct proc_dir_entry* usbwalldir = NULL;
static struct proc_dir_entry* usbwallstatus = NULL;
static struct proc_dir_entry* usbwallrelease = NULL;

static char status_buffer[USBWALL_PROC_STATUS_BUFFER_SIZE];
static char release_buffer[USBWALL_PROC_RELEASE_BUFFER_SIZE];


/*!
 ** \brief usbwall_status_read
 ** 
 ** \param page the kernel mapped memory page in which the data should be copied
 ** \param start unused
 ** \param off the current string offset asked by the userspace (if read in multiple times)
 ** \param count the asked bytes count to read
 ** \param eof EOF signal pointer to be setted
 ** \param data unused
 ** 
 ** \return the number of effectively copied bytes
 */
int usbwall_status_read(char *page,
                        char **start,
                        off_t off,
                        int count,
                        int *eof,
                        void *data)
{
   int is_eof = 0;

   DBG_TRACE(DBG_LEVEL_DEBUG, "entering status read, asking for reading %d bytes. Current offset is %ld", count, off);
   /*
   ** test buffer by now
   */
   sprintf(status_buffer, "USBWall module release %s\n", USBWALL_MODVERSION);
   print_keylist(status_buffer);
   if (off >= strlen(status_buffer)) {
     DBG_TRACE(DBG_LEVEL_WARNING, "offset greater than string size, leaving");
     count = 0;
     is_eof = 1;
     goto end;
   }
   if (count > (strlen(status_buffer) - off)) {
     count = (strlen(status_buffer) - off + 1);
     is_eof = 1;
   }
   snprintf(&(page[off]), count, &(status_buffer[off]));

end:
   if (is_eof) {
     DBG_TRACE(DBG_LEVEL_DEBUG, "EOF=1");
     /* update file offset */
      *eof = 1;
   } else {
      *eof = 0;
   }
   DBG_TRACE(DBG_LEVEL_DEBUG, "end of function. Returning %d", count);
   return count;
}

/*!
 ** \brief usbwall_release_read
 **
 ** Return the module release identifier, in order to be checked by the libusbwall to guarantee the interoperability
 ** 
 ** \param page the kernel mapped memory page in which the data should be copied
 ** \param start unused
 ** \param off the current string offset asked by the userspace (if read in multiple times)
 ** \param count the asked bytes count to read
 ** \param eof EOF signal pointer to be setted
 ** \param data unused
 ** 
 ** \return the number of effectively copied bytes
 */
int usbwall_release_read(char *page,
                         char **start,
                         off_t off,
                         int count,
                         int *eof,
                         void *data)
{
   int is_eof = 0;

   DBG_TRACE(DBG_LEVEL_DEBUG, "entering release read, asking for reading %d bytes. Current offset is %ld", count, off);
   /*
   ** return the release code
   */
   sprintf(release_buffer, "%d", USBWALL_MAJOR << 16 | USBWALL_MEDIUM << 8 | USBWALL_CURRENT);
   if (off >= strlen(release_buffer)) {
     DBG_TRACE(DBG_LEVEL_WARNING, "offset greater than string size, leaving");
     count = 0;
     is_eof = 1;
     goto end;
   }
   if (count > (strlen(release_buffer) - off)) {
     count = (strlen(release_buffer) - off + 1);
     is_eof = 1;
   }
   snprintf(&(page[off]), count, &(release_buffer[off]));

end:
   if (is_eof) {
     DBG_TRACE(DBG_LEVEL_DEBUG, "EOF=1");
     /* update file offset */
      *eof = 1;
   } else {
      *eof = 0;
   }
   DBG_TRACE(DBG_LEVEL_DEBUG, "end of function. Returning %d", count);
   return count;
}


/*!
 ** \fn usbwall_proc_init initialize the usbwall procfs itnerface
 ** 
 ** \return 1 if init failed, or 0.
 */
int usbwall_proc_init()
{
    usbwalldir = proc_mkdir("usbwall", NULL);
    if (usbwalldir == NULL) {
	goto fail_proc_mkdir;
    }
    usbwallstatus = create_proc_entry("status", 0400, usbwalldir);
    if (usbwallstatus == NULL) {
	goto fail_proc_entry_2;
    }
    usbwallrelease = create_proc_entry("release", 0400, usbwalldir);
    if (usbwallrelease == NULL) {
	goto fail_proc_entry_3;
    }


    usbwallstatus->read_proc = usbwall_status_read;
    usbwallrelease->read_proc = usbwall_release_read;
    return 0;

/* failure management - std linux usage */
fail_proc_entry_3:
    remove_proc_entry("release", usbwalldir);
fail_proc_entry_2:
    remove_proc_entry("status", usbwalldir);
    remove_proc_entry("usbwall", NULL);
fail_proc_mkdir:
    return 1;
}

/*!
 ** \fn usbwall_proc_release release the usbwall procfs interface
 */
void usbwall_proc_release()
{
    remove_proc_entry("release", usbwalldir);
    remove_proc_entry("status", usbwalldir);
    remove_proc_entry("usbwall", NULL);
}
